File: 2024/day14/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <regex>
#include <set>
#include <algorithm>
#include <tuple>

const int WIDTH = 101;
const int HEIGHT = 103;

struct Robot {
    int px, py, vx, vy;
};

std::vector<Robot> parseRobots(const std::string& text) {
    std::vector<Robot> robots;
    std::regex pattern(R"(p=(-?\d+),(-?\d+) v=(-?\d+),(-?\d+))");
    std::smatch match;

    std::istringstream stream(text);
    std::string line;

    while (std::getline(stream, line)) {
        if (std::regex_search(line, match, pattern)) {
            Robot r;
            r.px = std::stoi(match[1]);
            r.py = std::stoi(match[2]);
            r.vx = std::stoi(match[3]);
            r.vy = std::stoi(match[4]);
            robots.push_back(r);
        }
    }

    return robots;
}

std::vector<std::pair<int, int>> simulate(const std::vector<Robot>& robots, int seconds) {
    std::vector<std::pair<int, int>> positions;

    for (const auto& robot : robots) {
        // Position after 'seconds' time, with wrapping
        int new_x = ((robot.px + robot.vx * seconds) % WIDTH + WIDTH) % WIDTH;
        int new_y = ((robot.py + robot.vy * seconds) % HEIGHT + HEIGHT) % HEIGHT;
        positions.push_back({new_x, new_y});
    }

    return positions;
}

std::tuple<int, int, int, int> countQuadrants(const std::vector<std::pair<int, int>>& positions) {
    int mid_x = WIDTH / 2;   // 50
    int mid_y = HEIGHT / 2;  // 51

    int q1 = 0, q2 = 0, q3 = 0, q4 = 0;

    for (const auto& [x, y] : positions) {
        if (x == mid_x || y == mid_y) {
            continue;  // Skip robots on middle lines
        }

        if (x < mid_x && y < mid_y) {
            q1++;  // Top-left
        } else if (x > mid_x && y < mid_y) {
            q2++;  // Top-right
        } else if (x < mid_x && y > mid_y) {
            q3++;  // Bottom-left
        } else {
            q4++;  // Bottom-right
        }
    }

    return {q1, q2, q3, q4};
}

long long part1(const std::vector<Robot>& robots) {
    auto positions = simulate(robots, 100);
    auto [q1, q2, q3, q4] = countQuadrants(positions);
    return static_cast<long long>(q1) * q2 * q3 * q4;
}

int part2(const std::vector<Robot>& robots) {
    // The Christmas tree appears when robots cluster together
    // Look for a frame with a long horizontal line of robots (tree base/border)
    for (int seconds = 1; seconds <= WIDTH * HEIGHT; seconds++) {
        auto positions = simulate(robots, seconds);
        std::set<std::pair<int, int>> pos_set(positions.begin(), positions.end());

        // Look for a horizontal line of at least 20 consecutive robots
        for (int y = 0; y < HEIGHT; y++) {
            int max_consecutive = 0;
            int consecutive = 0;

            for (int x = 0; x < WIDTH; x++) {
                if (pos_set.count({x, y})) {
                    consecutive++;
                    max_consecutive = std::max(max_consecutive, consecutive);
                } else {
                    consecutive = 0;
                }
            }

            if (max_consecutive >= 20) {
                return seconds;
            }
        }
    }

    return -1;
}

int main() {
    // Read input file
    std::ifstream file("../input.txt");
    std::string input((std::istreambuf_iterator<char>(file)),
                      std::istreambuf_iterator<char>());

    // Parse robots
    auto robots = parseRobots(input);

    // Part 1
    std::cout << "Part 1: " << part1(robots) << std::endl;

    // Part 2
    std::cout << "Part 2: " << part2(robots) << std::endl;

    return 0;
}

File: 2024/day14/rust/solution.rs
use std::collections::HashSet;
use std::fs;

const WIDTH: i32 = 101;
const HEIGHT: i32 = 103;

#[derive(Debug, Clone, Copy)]
struct Robot {
    px: i32,
    py: i32,
    vx: i32,
    vy: i32,
}

fn parse_robots(text: &str) -> Vec<Robot> {
    let mut robots = Vec::new();

    for line in text.lines() {
        // Parse format: p=x,y v=vx,vy
        if let Some(p_start) = line.find("p=") {
            if let Some(v_start) = line.find(" v=") {
                let p_part = &line[p_start + 2..v_start];
                let v_part = &line[v_start + 3..];

                if let Some(comma_p) = p_part.find(',') {
                    if let Some(comma_v) = v_part.find(',') {
                        let px = p_part[..comma_p].parse::<i32>().ok();
                        let py = p_part[comma_p + 1..].parse::<i32>().ok();
                        let vx = v_part[..comma_v].parse::<i32>().ok();
                        let vy = v_part[comma_v + 1..].parse::<i32>().ok();

                        if let (Some(px), Some(py), Some(vx), Some(vy)) = (px, py, vx, vy) {
                            robots.push(Robot { px, py, vx, vy });
                        }
                    }
                }
            }
        }
    }

    robots
}

fn simulate(robots: &[Robot], seconds: i32) -> Vec<(i32, i32)> {
    robots
        .iter()
        .map(|robot| {
            // Use rem_euclid for proper modular arithmetic with negative numbers
            let new_x = (robot.px + robot.vx * seconds).rem_euclid(WIDTH);
            let new_y = (robot.py + robot.vy * seconds).rem_euclid(HEIGHT);
            (new_x, new_y)
        })
        .collect()
}

fn count_quadrants(positions: &[(i32, i32)]) -> (i32, i32, i32, i32) {
    let mid_x = WIDTH / 2;   // 50
    let mid_y = HEIGHT / 2;  // 51

    let mut q1 = 0;
    let mut q2 = 0;
    let mut q3 = 0;
    let mut q4 = 0;

    for &(x, y) in positions {
        // Skip robots on middle lines
        if x == mid_x || y == mid_y {
            continue;
        }

        if x < mid_x && y < mid_y {
            q1 += 1;  // Top-left
        } else if x > mid_x && y < mid_y {
            q2 += 1;  // Top-right
        } else if x < mid_x && y > mid_y {
            q3 += 1;  // Bottom-left
        } else {
            q4 += 1;  // Bottom-right
        }
    }

    (q1, q2, q3, q4)
}

fn part1(robots: &[Robot]) -> i32 {
    let positions = simulate(robots, 100);
    let (q1, q2, q3, q4) = count_quadrants(&positions);
    q1 * q2 * q3 * q4
}

fn part2(robots: &[Robot]) -> i32 {
    // The Christmas tree appears when robots cluster together
    // Look for a frame with a long horizontal line of robots (tree base/border)
    for seconds in 1..=(WIDTH * HEIGHT) {
        let positions = simulate(robots, seconds);
        let pos_set: HashSet<(i32, i32)> = positions.into_iter().collect();

        // Look for a horizontal line of at least 20 consecutive robots
        for y in 0..HEIGHT {
            let mut max_consecutive = 0;
            let mut consecutive = 0;

            for x in 0..WIDTH {
                if pos_set.contains(&(x, y)) {
                    consecutive += 1;
                    max_consecutive = max_consecutive.max(consecutive);
                } else {
                    consecutive = 0;
                }
            }

            if max_consecutive >= 20 {
                return seconds;
            }
        }
    }

    -1
}

fn main() {
    let input_text = fs::read_to_string("../input.txt")
        .expect("Failed to read input.txt");

    let robots = parse_robots(&input_text);

    println!("Part 1: {}", part1(&robots));
    println!("Part 2: {}", part2(&robots));
}
